[JIT] Trace module calls

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28474 [JIT] Trace module calls**
* #28864 [JIT] API for finding a common ancestor block for a pair of nodes

Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18273353](https://our.internmc.facebook.com/intern/diff/D18273353)

Commit:
[WIP][JIT] Trace module calls

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"




[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+  // Assign each block a number, with graph->block() gettng
+  // 0 and each sub-block gets a higher number based on how
+  // deeply nested it is.
+  BlockInfoMap block_info;
```
We should put level information into the Block itself. It does not change once a block is created so it is just `owningNode()->block()->level() + 1`

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+    Block* b,
+    RemappingTable& remap,
+    const BlockInfoMap& block_info) {
+  auto find_common_ancestor_block_level = [&](Node* n) {
```
I think this should also be an API on Node, and can be more easily implemented when the level of each node is known pseudo code:

`while (one level is deeper than another) { go up one level }; while (block_lhs != block_rhs) { make both nodes go up a level; }

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+
+void makeDefsDominateUses(
+    Block* b,
+    RemappingTable& remap,
```
Needs a comment about what this holds.

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+        int ancestor_level = find_common_ancestor_block_level(inp->node());
+
+        Block* origin_block = inp->node()->owningBlock();
+        if (ancestor_level == block_info.at(origin_block)) {
```
I don't think this special case needs to exist. It seems like it is just the degenerate version of the code below.

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+          Block* b = origin_block;
+
+          // Start off where we left off if there's an existing mapping
+          if (remap.count(inp)) {
```
Because I don't know what remap holds (e.g. what does the level mean), I am finding it hard to follow this logic.

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+};
+using RemappingTable = std::unordered_map<Value*, RemappedValueInfo>;
+
+void makeDefsDominateUses(
```
This transform might be easier to follow as a class because it already contains lambda functions, and there is a several shared members (remap, block_info).

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+        Value* tup_output = n->addOutput()->setType(TupleType::create(types));
+        Node* tup_unpack = g->createTupleUnpack(tup_output)->insertAfter(n);
+        for (size_t i = 0; i < tup_unpack->outputs().size(); ++i) {
+          n->output(0)->replaceAllUsesWith(tup_unpack->output(i));
```
This is n^2 to erase the first output each time.

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+    }
+  }
+
+  // Inline base __module call
```
I don't understand what is going on here. Seems like something that should be handled separately. Comment explains what but not why

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+
+namespace {
+
+std::vector<std::string> splitModuleQualname(const std::string& s) {
```
QualifiedName already implements this.

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+
+      Function* f = nullptr;
+
+      // Find a unique name to add this method as
```
Please make naming mangling into a function.

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+    std::vector<std::string> prefix) {
+  Value* self;
+  if (g->inputs().size() == 0 ||
+      g->inputs()[0]->type() != this_level_self.type()) {
```
I don't know what this is suppose to be checking for, but pointer comparison of the types is most likely wrong.

zdevito replied:
Something feels wrong here. There is no guarantee that something with the same type as self is the self module (we should not assume they are singletons).  I feel like Module self should just be the first argument everywhere so it doesn't have to be recovered here.

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+    // class compilation unit and register that Function as a method
+    // on the corresponding Module in the Module hierarchy. Note that we
+    // unique the methods by naming them forward, forward1, forward2...
+    createMethodCalls(graph, *self, {"__module"});
```
Can you make an argument for the algorithmic complexity of this whole process? I want to make sure it is not N^2 as the module depth grows, but I cannot tell by inspection. Most of these passes are recursive so it is not clear what happens when we repeatedly apply them.

zdevito commented on torch/jit/__init__.py:
```diff
@@ -280,7 +269,10 @@ def get_trace_graph(f, args=(), kwargs=None, _force_outplace=False,
     if not isinstance(args, tuple):
         args = (args,)
-    return ONNXTracedModule(f, _force_outplace, return_inputs, _return_inputs_states)(*args, **kwargs)
+    torch.jit._trace_module_map = {}
```
You get _one_ place where this global is set/reset.  It is really hard to know where it has to be manipulated at the moment because it exists in three different places. Pass it in as an argument up until it is no longer feasibly to do so.

zdevito commented on torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp:
```diff
@@ -0,0 +1,350 @@
+
+struct RemappedValueInfo {
+  Value* v;
+  int level;
```
not sure why this needs to be tracked.

zdevito reviewed (commented):
This is looking pretty good. My overall comment is that is pretty hard to follow the logic in the pass due to some missing comments. Putting the find common node logic in ir.cpp would help, as would making makiDefsDominateUses into a  class.

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

zdevito commented on torch/nn/modules/module.py:
```diff
@@ -500,32 +501,30 @@ def register_forward_hook(self, hook):
+            if name:
+                cur_scope_name = tracing_state.curr_scope()
+                if not name.startswith(cur_scope_name):
+                    warnings.warn('Called module {} from {}, but these modules are not in a '
```
This restriction isn't necessary. The nesting of ScopeBlocks does not have to match the nesting of the module hierarchy. In the cases where it doesn't the module whose method _is_ being called is going to need to be passed in as a non-self argument to calling method. But this is something that is already handled by the capture process. Essentially this can be done with two steps:

1. Before converting subblocks, just add the correct `self` argument to the block, which is determined by the _trace_module_map.
2. Also before converting subblocks, replace all uses of variables inside the immediate subblock of values that are actually attributes of self for that subblock with their getattr equivalents. This would make sure we do not pass self.weight as an argument to the method but look it up from the module. This will also make all those `self` arguments we added in step 1 into get attrs on the parent when possible.

In cases where we do not have a hierarchy, then modules will get passed around as arguments that are not self, but the code will otherwise be correct.

zdevito reviewed (commented):
Haven't finished reading the actual transform code, but see the one comment so far below.

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18121008](https://our.internmc.facebook.com/intern/diff/D18121008)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18273353](https://our.internmc.facebook.com/intern/diff/D18273353)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18273353](https://our.internmc.facebook.com/intern/diff/D18273353)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18273353](https://our.internmc.facebook.com/intern/diff/D18273353)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace module calls"


Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3


TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18273353](https://our.internmc.facebook.com/intern/diff/D18273353)

[ghstack-poisoned]

jamesr66a commented:
Closing because ghstack fuckery
